### PR TITLE
Resource api

### PR DESCRIPTION
**Pull request by leplatrem**
While writing the docs, I realized too much stuff was public in resource class

### CODE REVIEW

**Comment by leplatrem** [comment]
@ametaireau @Natim r?

**Comment by Natim** [comment]
Le retour des underscores :)

File: cliquet/resource.py [hunk 1]
```diff
@@ -120,30 +100,36 @@ def fetch_record(self):
...
-                self.raise_invalid(name=field, description=error)
+                self._raise_invalid(name=field, description=error)
 
         updated = record.copy()
         updated.update(**changes)
-        return self.validate(updated)
+        return self._validate(updated)
 
-    def validate(self, record):
+    def _validate(self, record):
```
Comment (ametaireau) [inline cliquet/resource.py]: Should this stay public, in case a library user would like to define its own validation logic?

**Comment by ametaireau** [comment]
I'm a little fuzzy about what we're using private / public for here.

Since this is a base class, I read it like that: from inside the class, everything prefixed with an underscore is to be accessed by the base class, and not by the subclasses.

If this is the case, I believe the helpers to send errors should be public, since someone defining a subclass could make good use of such helpers (and I believe if they see them they will start to use them anyway, even if they start with an underscore, so we better put them public from the start).

**Comment by Natim** [comment]
> they will start to use them anyway, even if they start with an underscore, so we better put them public from the start

Exactly my point.

File: cliquet/resource.py [hunk 2]
```diff
@@ -154,13 +140,13 @@ def validate(self, record):
                 self.request.errors.add('body', name=field, description=error)
             raise errors.json_error(self.request.errors)
 
-    def add_timestamp_header(self, response):
+    def _add_timestamp_header(self, response):
```
Comment (Natim) [inline cliquet/resource.py]: Should this stay public, in case a library user would like to add it to other endpoints?

File: cliquet/resource.py [hunk 3]
```diff
@@ -154,13 +140,13 @@ def validate(self, record):
...
         """Add current timestamp in response headers, when request comes in.
         """
         timestamp = six.text_type(self.timestamp).encode('utf-8')
         response.headers['Last-Modified'] = timestamp
 
-    def raise_304_if_not_modified(self, record=None):
+    def _raise_304_if_not_modified(self, record=None):
```
Comment (Natim) [inline cliquet/resource.py]: Should this stay public, in case a library user would like to add it to other endpoints?

File: cliquet/resource.py [hunk 4]
```diff
@@ -176,10 +162,10 @@ def raise_304_if_not_modified(self, record=None):
 
             if current_timestamp <= modified_since:
                 response = HTTPNotModified()
-                self.add_timestamp_header(response)
+                self._add_timestamp_header(response)
                 raise response
 
-    def raise_412_if_modified(self, record=None):
+    def _raise_412_if_modified(self, record=None):
```
Comment (Natim) [inline cliquet/resource.py]: Should this stay public, in case a library user would like to add it to other endpoints?

File: cliquet/resource.py [hunk 5]
```diff
@@ -202,10 +188,16 @@ def raise_412_if_modified(self, record=None):
                         error=HTTPPreconditionFailed.title,
                         message=error_msg),
                     content_type='application/json')
-                self.add_timestamp_header(response)
+                self._add_timestamp_header(response)
                 raise response
 
-    def raise_conflict(self, exception):
+    def _raise_invalid(self, location='body', **kwargs):
```
Comment (Natim) [inline cliquet/resource.py]: Should this stay public, in case a library user would like to add it to other endpoints?

File: cliquet/resource.py [hunk 6]
```diff
@@ -202,10 +188,16 @@ def raise_412_if_modified(self, record=None):
...
+        """Helper to raise a validation error."""
+        self.request.errors.add(location, **kwargs)
+        response = errors.json_error(self.request.errors)
+        raise response
+
+    def _raise_conflict(self, exception):
```
Comment (Natim) [inline cliquet/resource.py]: Should this stay public, in case a library user would like to add it to other endpoints?

**Comment by Natim** [comment]
I am ok with the other changes.

**Comment by ametaireau** [comment]
Thanks for updating the pull request, @Natim. I'm okay to sign-off this merge, but since @leplatrem started it I'll let him review and merge.

**Comment by Natim** [comment]
This apparently needs a rebase.

**Pull request merged**